core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -145,18 +145,20 @@ abstract class AbstractFetcherThread(name: String,
     val partitionsWithoutEpochs = mutable.Set.empty[TopicPartition]
     val partitionsWithEpochs = mutable.Map.empty[TopicPartition, EpochData]
 
-    partitionStates.partitionStates.asScala.foreach { state =>
-      val tp = state.topicPartition
-      if (state.value.isTruncating) {
-        latestEpoch(tp) match {
-          case Some(latestEpoch) =>
-            val partitionData = new EpochData(Optional.of(state.value.currentLeaderEpoch), latestEpoch)
-            partitionsWithEpochs += tp -> partitionData
-          case None =>
-            partitionsWithoutEpochs += tp
+    partitionStates.stream().forEach(new Consumer[PartitionStates.PartitionState[PartitionFetchState]] {
```

Comment:
Not sure if this part of the change was worth it. We still need a copy per element and the code is a bit harder to read. The rest looks great though.